# Global Weather Data Analysis
- this is an analysis project to understand the weather of countries in 2024

In [645]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

- pandas for data manipulation
- matplotlib and seaborn for visualization

## Loading Data

In [646]:
df=pd.read_csv('weather.csv')
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


## Inspecting Data

In [647]:
df.head()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55


In [648]:
df.tail()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9
136632,Zimbabwe,Harare,-17.8178,31.0447,Africa/Harare,1776666600,2026-04-20 08:30,20.0,68.1,Sunny,...,7.85,8.05,1,1,06:06,17:43,09:07,20:18,Waxing Crescent,9


In [649]:
df.sample()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
88040,Slovakia,Bratislava,48.15,17.1167,Europe/Bratislava,1754986500,2025-08-12 10:15,22.2,72.0,Sunny,...,12.765,16.465,1,2,05:42,20:10,21:28,09:19,Waning Gibbous,92


## Checking Data Types

In [650]:
df.dtypes
# check column data type

country                          object
location_name                    object
latitude                        float64
longitude                       float64
timezone                         object
last_updated_epoch                int64
last_updated                     object
temperature_celsius             float64
temperature_fahrenheit          float64
condition_text                   object
wind_mph                        float64
wind_kph                        float64
wind_degree                       int64
wind_direction                   object
pressure_mb                       int64
pressure_in                     float64
precip_mm                       float64
precip_in                       float64
humidity                          int64
cloud                             int64
feels_like_celsius              float64
feels_like_fahrenheit           float64
visibility_km                   float64
visibility_miles                  int64
uv_index                        float64


In [651]:
df.columns
#Display column names

Index(['country', 'location_name', 'latitude', 'longitude', 'timezone',
       'last_updated_epoch', 'last_updated', 'temperature_celsius',
       'temperature_fahrenheit', 'condition_text', 'wind_mph', 'wind_kph',
       'wind_degree', 'wind_direction', 'pressure_mb', 'pressure_in',
       'precip_mm', 'precip_in', 'humidity', 'cloud', 'feels_like_celsius',
       'feels_like_fahrenheit', 'visibility_km', 'visibility_miles',
       'uv_index', 'gust_mph', 'gust_kph', 'air_quality_Carbon_Monoxide',
       'air_quality_Ozone', 'air_quality_Nitrogen_dioxide',
       'air_quality_Sulphur_dioxide', 'air_quality_PM2.5', 'air_quality_PM10',
       'air_quality_us-epa-index', 'air_quality_gb-defra-index', 'sunrise',
       'sunset', 'moonrise', 'moonset', 'moon_phase', 'moon_illumination'],
      dtype='object')

In [652]:
df.shape
# display number or rows and columns

(136633, 41)

In [653]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   country                       136633 non-null  object 
 1   location_name                 136633 non-null  object 
 2   latitude                      136633 non-null  float64
 3   longitude                     136633 non-null  float64
 4   timezone                      136633 non-null  object 
 5   last_updated_epoch            136633 non-null  int64  
 6   last_updated                  136633 non-null  object 
 7   temperature_celsius           136633 non-null  float64
 8   temperature_fahrenheit        136633 non-null  float64
 9   condition_text                136633 non-null  object 
 10  wind_mph                      136633 non-null  float64
 11  wind_kph                      136633 non-null  float64
 12  wind_degree                   136633 non-nul

### Dataset Overview
- number of rows = 136633
- number of columns = 41
- Data Types : float = 21, integer = 9, objects = 11
- missing values =none 
- sunrise, moonset, moonrise and sunset times stored as object
- last_updated stored as objects             
- memory used = 42.7 + MB

## Selecting Columns
- selecting single columns 

In [654]:
df['country']

0         Afghanistan
1             Albania
2             Algeria
3             Andorra
4              Angola
             ...     
136628      Venezuela
136629        Vietnam
136630          Yemen
136631         Zambia
136632       Zimbabwe
Name: country, Length: 136633, dtype: object

- selecting multipple columns

In [655]:
df[['sunrise','sunset']]

,sunrise,sunset
0,04:50,18:50
1,05:21,19:54
2,05:40,19:50
3,06:31,21:11
4,06:12,17:55
...,...,...
136628,06:14,18:39
136629,05:34,18:17
136630,05:46,18:18
136631,06:15,17:56


In [656]:
place = df[['country','location_name']]

In [657]:
place

,country,location_name
0,Afghanistan,Kabul
1,Albania,Tirana
2,Algeria,Algiers
3,Andorra,Andorra La Vella
4,Angola,Luanda
...,...,...
136628,Venezuela,Caracas
136629,Vietnam,Hanoi
136630,Yemen,Sanaa
136631,Zambia,Lusaka


## Filtering The Data

### 1

In [658]:
df[df['temperature_celsius']==25]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
31,Cameroon,Yaounde,3.8700,11.5200,Africa/Douala,1715849100,2024-05-16 09:45,25.0,76.9,Partly Cloudy,...,25.60,31.20,2,3,06:01,18:19,13:13,00:59,Waxing Gibbous,55
43,Cyprus,Nicosia,35.1700,33.3700,Asia/Famagusta,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,7.70,13.50,1,1,05:42,19:45,13:11,02:08,Waxing Gibbous,55
82,Israel,Jerusalem,31.7800,35.2300,Asia/Jerusalem,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,5.00,9.40,1,1,05:42,19:30,13:07,01:55,Waxing Gibbous,55
144,Rwanda,Kigali,-1.9500,30.0600,Africa/Kigali,1715849100,2024-05-16 10:45,25.0,77.0,Partly cloudy,...,4.20,6.30,1,1,05:55,17:57,13:02,00:36,Waxing Gibbous,55
172,Tanzania,Dodoma,-6.1800,35.7500,Africa/Dar_es_Salaam,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,1.90,3.40,1,1,06:38,18:28,13:43,01:08,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136101,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776493800,2026-04-18 07:30,25.0,77.0,Patchy light rain with thunder,...,6.65,9.75,1,1,06:18,18:30,06:54,19:33,Waxing Crescent,0
136389,Saint Lucia,Castries,14.0000,-61.0000,America/St_Lucia,1776579300,2026-04-19 02:15,25.0,77.0,Partly cloudy,...,8.55,13.35,1,1,05:48,18:18,07:22,20:48,Waxing Crescent,4
136390,Saint Vincent and the Grenadines,Kingstown,13.1333,-61.2167,America/St_Vincent,1776579300,2026-04-19 02:15,25.0,77.0,Partly cloudy,...,6.60,10.00,1,1,05:50,18:18,07:25,20:47,Waxing Crescent,4
136483,Democratic Republic of Congo,Kinshasa,-4.3000,15.3000,Africa/Kinshasa,1776666600,2026-04-20 07:30,25.0,77.1,Patchy rain nearby,...,14.75,14.95,1,2,05:58,17:58,08:42,20:55,Waxing Crescent,9


### 2

In [659]:
df[df['moon_phase']=='New Moon']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4285,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1717682400,2024-06-06 18:30,18.0,64.3,Patchy rain nearby,...,3.90,8.20,1,1,04:40,19:04,04:02,19:18,New Moon,0
4286,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1717682400,2024-06-06 16:00,32.0,89.6,Sunny,...,4.50,5.10,1,1,05:08,20:11,04:31,20:41,New Moon,0
4287,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1717682400,2024-06-06 15:00,32.0,89.6,Sunny,...,16.40,34.80,2,2,05:29,20:04,04:58,20:33,New Moon,0
4288,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1717682400,2024-06-06 16:00,22.5,72.6,Patchy rain nearby,...,2.20,3.40,1,1,06:17,21:29,05:43,22:03,New Moon,0
4289,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1717682400,2024-06-06 15:00,26.0,78.8,Partly cloudy,...,7.30,25.20,1,1,06:17,17:55,06:04,17:55,New Moon,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136196,Samoa,Apia,-13.8333,-171.7333,Pacific/Apia,1776493800,2026-04-18 19:30,27.4,81.3,Partly cloudy,...,4.65,6.75,1,1,06:33,18:19,06:50,18:40,New Moon,0
136207,Solomon Islands,Honiara,-9.4333,159.9500,Pacific/Guadalcanal,1776493800,2026-04-18 17:30,31.2,88.2,Patchy rain nearby,...,2.45,2.95,1,1,06:23,18:16,06:42,18:44,New Moon,0
136224,Tonga,Nuku`Aloia,-21.1333,-175.2000,Pacific/Tongatapu,1776493800,2026-04-18 19:30,24.1,75.4,Clear,...,6.15,9.45,1,1,06:53,18:27,07:14,18:43,New Moon,0
136229,Tuvalu,Funafuti,-8.5167,179.2167,Pacific/Funafuti,1776493800,2026-04-18 18:30,30.3,86.5,Partly cloudy,...,1.75,2.25,1,1,06:06,17:59,06:21,18:25,New Moon,0


### 3

In [660]:
df[df['condition_text']=='Patchy light rain with thunder']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
24,Brunei Darussalam,Bandar Seri Begawan,4.8800,114.9300,Asia/Brunei,1715849100,2024-05-16 16:45,32.0,89.6,Patchy light rain with thunder,...,2.20,2.30,1,1,06:06,18:27,13:06,00:54,Waxing Gibbous,55
97,Thailand,Nan,18.7800,100.7800,Asia/Bangkok,1715849100,2024-05-16 15:45,33.0,91.4,Patchy light rain with thunder,...,2.80,3.00,1,1,05:43,18:44,12:50,01:10,Waxing Gibbous,55
219,Brunei Darussalam,Bandar Seri Begawan,4.8800,114.9300,Asia/Brunei,1715868000,2024-05-16 22:00,27.0,80.6,Patchy light rain with thunder,...,9.50,9.80,1,1,06:06,18:27,13:06,00:54,Waxing Gibbous,55
234,Congo,Brazzaville,-4.2600,15.2800,Africa/Brazzaville,1715868000,2024-05-16 15:00,32.0,89.6,Patchy light rain with thunder,...,15.00,31.10,1,2,05:58,17:53,13:05,00:34,Waxing Gibbous,55
240,Democratic Republic of Congo,Kinshasa,-4.3000,15.3000,Africa/Brazzaville,1715868000,2024-05-16 15:00,32.0,89.6,Patchy light rain with thunder,...,15.00,31.10,1,2,05:58,17:53,13:05,00:34,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136101,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776493800,2026-04-18 07:30,25.0,77.0,Patchy light rain with thunder,...,6.65,9.75,1,1,06:18,18:30,06:54,19:33,Waxing Crescent,0
136274,Cameroon,Douala,4.0503,9.7000,Africa/Douala,1776579300,2026-04-19 07:15,27.4,81.3,Patchy light rain with thunder,...,32.05,35.15,2,3,06:14,18:27,07:47,20:32,Waxing Crescent,3
136296,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776579300,2026-04-19 07:15,27.3,81.1,Patchy light rain with thunder,...,15.85,19.45,2,2,06:17,18:30,07:51,20:35,Waxing Crescent,3
136469,Cameroon,Douala,4.0503,9.7000,Africa/Douala,1776666600,2026-04-20 07:30,24.1,75.4,Patchy light rain with thunder,...,27.85,31.95,2,3,06:13,18:27,08:48,21:37,Waxing Crescent,9


### 4

In [661]:
df[df['moon_illumination']==0]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4285,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1717682400,2024-06-06 18:30,18.0,64.3,Patchy rain nearby,...,3.90,8.20,1,1,04:40,19:04,04:02,19:18,New Moon,0
4286,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1717682400,2024-06-06 16:00,32.0,89.6,Sunny,...,4.50,5.10,1,1,05:08,20:11,04:31,20:41,New Moon,0
4287,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1717682400,2024-06-06 15:00,32.0,89.6,Sunny,...,16.40,34.80,2,2,05:29,20:04,04:58,20:33,New Moon,0
4288,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1717682400,2024-06-06 16:00,22.5,72.6,Patchy rain nearby,...,2.20,3.40,1,1,06:17,21:29,05:43,22:03,New Moon,0
4289,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1717682400,2024-06-06 15:00,26.0,78.8,Partly cloudy,...,7.30,25.20,1,1,06:17,17:55,06:04,17:55,New Moon,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136237,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1776493800,2026-04-18 17:30,26.0,78.8,Light rain shower,...,5.55,8.45,1,1,05:57,17:36,06:18,17:57,New Moon,0
136239,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776493800,2026-04-18 13:30,27.3,81.1,Partly cloudy,...,106.35,109.55,4,10,05:36,18:16,05:48,19:16,Waxing Crescent,0
136240,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776493800,2026-04-18 09:30,21.4,70.5,Sunny,...,22.05,114.75,2,2,05:47,18:18,06:10,19:24,Waxing Crescent,0
136241,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776493800,2026-04-18 08:30,20.6,69.0,Partly Cloudy,...,5.65,5.75,1,1,06:15,17:58,07:00,18:40,Waxing Crescent,0


### 5

In [662]:
df[(df['temperature_celsius']==20)&(df['condition_text']=='Sunny')]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
119,Namibia,Windhoek,-22.5700,17.0800,Africa/Windhoek,1715849100,2024-05-16 10:45,20.0,68.0,Sunny,...,2.10,5.50,1,1,07:18,18:18,14:14,01:06,Waxing Gibbous,55
161,South Africa,Pretoria,-25.7500,28.1900,Africa/Johannesburg,1715849100,2024-05-16 10:45,20.0,68.0,Sunny,...,78.10,108.30,4,10,06:38,17:29,13:32,00:16,Waxing Gibbous,55
168,Sweden,Stockholm,59.3300,18.0500,Europe/Stockholm,1715849100,2024-05-16 10:45,20.0,68.0,Sunny,...,4.20,6.10,1,1,04:12,21:18,12:22,03:11,Waxing Gibbous,55
250,Estonia,Tallinn,59.4300,24.7300,Europe/Tallinn,1715868000,2024-05-16 17:00,20.0,68.0,Sunny,...,8.00,8.40,1,1,04:44,21:52,12:54,03:45,Waxing Gibbous,55
337,Romania,Bucharest,44.4300,26.1000,Europe/Bucharest,1715868900,2024-05-16 17:15,20.0,68.0,Sunny,...,2.90,3.70,1,1,05:47,20:38,13:27,02:54,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
135359,Kyrghyzstan,Bishkek,42.8731,74.6003,Asia/Bishkek,1776150000,2026-04-14 13:00,20.0,68.0,Sunny,...,9.75,12.55,1,1,06:22,19:43,04:44,16:09,Waning Crescent,17
135376,Mauritania,Nouakchott,18.0864,-15.9753,Africa/Nouakchott,1776150000,2026-04-14 07:00,20.0,68.0,Sunny,...,45.25,236.95,3,5,06:48,19:21,04:33,16:39,Waning Crescent,15
135543,Iraq,Baghdad,33.3386,44.3939,Asia/Baghdad,1776234600,2026-04-15 09:30,20.0,68.1,Sunny,...,8.85,13.35,1,1,05:33,18:33,04:05,16:26,Waning Crescent,9
136046,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776407400,2026-04-17 08:30,20.0,67.9,Sunny,...,5.05,5.35,1,1,06:14,17:58,05:58,17:52,New Moon,1


## Selecting Rows
- df.iloc()
- df.iloc()

In [663]:
df.iloc[4]
# single row

country                                   Angola
location_name                             Luanda
latitude                                   -8.84
longitude                                  13.23
timezone                           Africa/Luanda
last_updated_epoch                    1715849100
last_updated                    2024-05-16 09:45
temperature_celsius                         26.0
temperature_fahrenheit                      78.8
condition_text                     Partly cloudy
wind_mph                                     8.1
wind_kph                                    13.0
wind_degree                                  150
wind_direction                               SSE
pressure_mb                                 1011
pressure_in                                29.85
precip_mm                                    0.0
precip_in                                    0.0
humidity                                      89
cloud                                         50
feels_like_celsius  

In [664]:
df.iloc[1:4]
#portions

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55


### loc methord

In [665]:
df.loc[100,'sunrise']

'05:49'

## Rename Columns
- df.rename(columns={'old_name',:'new name},inplace=True)

In [666]:
df.rename(columns={'location_name':'location','condition_set':'condition'},inplace=True)

In [667]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [668]:
df.rename(columns={'air_quality_us-epa-index':'epa_aqi','air_quality_gb-defra-index':'daqi'},inplace=True)

In [669]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,06:14,18:39,08:54,22:12,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,05:34,18:17,07:30,21:34,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,05:46,18:18,08:00,21:39,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,06:15,17:56,09:13,20:35,Waxing Crescent,9


In [670]:
df['moon_phase'].value_counts()

moon_phase
Waning Crescent    29886
Waxing Crescent    29481
Waxing Gibbous     29382
Waning Gibbous     28477
Full Moon           4933
Last Quarter        4920
New Moon            4876
First Quarter       4678
Name: count, dtype: int64

# Data cleaning

### 5.2 Cleaning Numerical Columns With Charecters

In [671]:
df['last_updated']=pd.to_datetime(df['last_updated'],errors='coerce')

In [672]:
df.head()

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,04:50,18:50,12:12,01:11,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,05:21,19:54,12:58,02:14,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,73.4,Sunny,...,10.4,18.4,1,1,05:40,19:50,13:15,02:14,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,06:31,21:11,14:12,03:31,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,06:12,17:55,13:17,00:38,Waxing Gibbous,55


In [673]:
df.dtypes

country                                 object
location                                object
latitude                               float64
longitude                              float64
timezone                                object
last_updated_epoch                       int64
last_updated                    datetime64[ns]
temperature_celsius                    float64
temperature_fahrenheit                 float64
condition_text                          object
wind_mph                               float64
wind_kph                               float64
wind_degree                              int64
wind_direction                          object
pressure_mb                              int64
pressure_in                            float64
precip_mm                              float64
precip_in                              float64
humidity                                 int64
cloud                                    int64
feels_like_celsius                     float64
feels_like_fa

- delete sunrise, sunset, moonrise, moonset, moon_phase, mmoon_illumination

## Deleting unwanted columns

In [674]:
df.drop(columns=['sunrise','sunset','moonrise','moonset','moon_phase','moon_illumination'],inplace=True)

In [675]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,gust_mph,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,9.5,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,...,11.4,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,73.4,Sunny,...,13.9,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,43.3,Light drizzle,...,8.5,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,...,12.5,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,63.6,Partly Cloudy,...,6.7,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,82.9,Cloudy,...,12.7,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,69.7,Sunny,...,5.7,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,70.8,Sunny,...,11.1,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [676]:
df.drop(columns=['temperature_fahrenheit','wind_mph','feels_like_fahrenheit','precip_in','visibility_miles','gust_mph','pressure_in'],inplace=True)

In [677]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,condition_text,wind_kph,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,...,7.0,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,...,5.0,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,Sunny,15.1,...,5.0,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,Light drizzle,11.9,...,2.0,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,...,8.0,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,Partly Cloudy,3.6,...,0.0,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,Cloudy,17.6,...,5.5,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,Sunny,7.9,...,6.5,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,Sunny,13.0,...,1.6,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [678]:
df.drop(columns=['last_updated_epoch'],inplace=True)

In [679]:
df

,country,location,latitude,longitude,timezone,last_updated,temperature_celsius,condition_text,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,320,...,5.0,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,2024-05-16 09:45:00,23.0,Sunny,15.1,280,...,5.0,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,2024-05-16 10:45:00,6.3,Light drizzle,11.9,215,...,2.0,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,150,...,8.0,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,2026-04-20 02:30:00,17.5,Partly Cloudy,3.6,122,...,0.0,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,2026-04-20 13:30:00,28.3,Cloudy,17.6,151,...,5.5,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,2026-04-20 09:30:00,21.0,Sunny,7.9,135,...,6.5,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,2026-04-20 08:30:00,21.5,Sunny,13.0,58,...,1.6,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [680]:
df.shape

(136633, 27)

## 27 columns frome 41

In [681]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 27 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  object        
 1   location                      136633 non-null  object        
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  object        
 5   last_updated                  136633 non-null  datetime64[ns]
 6   temperature_celsius           136633 non-null  float64       
 7   condition_text                136633 non-null  object        
 8   wind_kph                      136633 non-null  float64       
 9   wind_degree                   136633 non-null  int64         
 10  wind_direction                136633 non-null  object        
 11  pressure_mb  

## Handling Missing Values

In [682]:
df.isnull().sum()

country                         0
location                        0
latitude                        0
longitude                       0
timezone                        0
last_updated                    0
temperature_celsius             0
condition_text                  0
wind_kph                        0
wind_degree                     0
wind_direction                  0
pressure_mb                     0
precip_mm                       0
humidity                        0
cloud                           0
feels_like_celsius              0
visibility_km                   0
uv_index                        0
gust_kph                        0
air_quality_Carbon_Monoxide     0
air_quality_Ozone               0
air_quality_Nitrogen_dioxide    0
air_quality_Sulphur_dioxide     0
air_quality_PM2.5               0
air_quality_PM10                0
epa_aqi                         0
daqi                            0
dtype: int64

- find median and mode to fill in missing data
- #no missing data in this dataset

In [683]:
df['temperature_celsius'].median()

23.9

In [684]:
df['timezone'].mode()

0    Asia/Bangkok
Name: timezone, dtype: object

Finding mean,median and mode of each columns 

In [685]:
print(df.select_dtypes(include='number').mean())
# to find mean of numerical columns

latitude                          19.211173
longitude                         21.956700
temperature_celsius               21.268191
wind_kph                          12.897045
wind_degree                      168.797348
pressure_mb                     1014.064304
precip_mm                          0.134168
humidity                          66.661553
cloud                             39.976228
feels_like_celsius                22.120190
visibility_km                      9.507450
uv_index                           3.313816
gust_kph                          18.242712
air_quality_Carbon_Monoxide      459.169559
air_quality_Ozone                 57.997988
air_quality_Nitrogen_dioxide      14.974117
air_quality_Sulphur_dioxide       10.315061
air_quality_PM2.5                 24.149276
air_quality_PM10                  48.293201
epa_aqi                            1.697430
daqi                               2.601275
dtype: float64


In [686]:
df['latitude'].median()



17.25

In [687]:
print(df.select_dtypes(include='number').median())
#to find median of numerical columns

latitude                          17.2500
longitude                         23.2361
temperature_celsius               23.9000
wind_kph                          10.8000
wind_degree                      161.0000
pressure_mb                     1014.0000
precip_mm                          0.0000
humidity                          72.0000
cloud                             30.0000
feels_like_celsius                25.1000
visibility_km                     10.0000
uv_index                           1.9000
gust_kph                          15.3000
air_quality_Carbon_Monoxide      293.7000
air_quality_Ozone                 55.0000
air_quality_Nitrogen_dioxide       5.6500
air_quality_Sulphur_dioxide        2.4050
air_quality_PM2.5                 14.1500
air_quality_PM10                  19.9800
epa_aqi                            1.0000
daqi                               2.0000
dtype: float64


In [688]:
c=df.select_dtypes(include='object')
for i in c:
    print(i,df[i].mode())
    print()
    # to find mode of catogarical column

country 0    Bulgaria
Name: country, dtype: object

location 0            Accra
1            Amman
2           Asmara
3          Baghdad
4             Bern
5        Bujumbura
6            Dakar
7            Kabul
8             Kyiv
9           Malabo
10       N'djamena
11             Nan
12           Sanaa
13            Suva
14        Tashkent
15           Tokyo
16        Valletta
17    Vatican City
18          Warsaw
Name: location, dtype: object

timezone 0    Asia/Bangkok
Name: timezone, dtype: object

condition_text 0    Partly cloudy
Name: condition_text, dtype: object

wind_direction 0    E
Name: wind_direction, dtype: object



## Handling Duplicates

In [689]:
df.duplicated().sum()

np.int64(1)

In [690]:
print('before',df.shape)
df=df.drop_duplicates()
print('after',df.shape)

before (136633, 27)
after (136632, 27)


In [691]:
df.duplicated().sum()

np.int64(0)

In [692]:
df.sample(10)

,country,location,latitude,longitude,timezone,last_updated,temperature_celsius,condition_text,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
29948,Djibouti,Djibouti,11.5950,43.1481,Africa/Djibouti,2024-10-18 12:30:00,34.3,Sunny,11.9,79,...,10.4,13.7,234.95,91.0,2.775,4.995,40.145,109.150,2,4
120567,Papua New Guinea,Port Moresby,-9.4647,147.1925,Pacific/Port_Moresby,2026-01-27 17:00:00,29.2,Partly Cloudy,20.2,290,...,0.3,26.8,180.85,33.0,1.850,1.050,8.750,10.050,1,1
53062,El Salvador,San Salvador,13.7086,-89.2031,America/El_Salvador,2025-02-14 04:00:00,25.1,Partly cloudy,4.3,357,...,0.0,8.4,425.50,124.0,2.220,10.730,20.720,22.200,2,2
126977,Micronesia,Palikir,6.9167,158.1500,Pacific/Pohnpei,2026-03-01 17:30:00,28.1,Overcast,7.2,97,...,1.2,12.6,98.85,36.0,0.950,0.950,4.550,6.850,1,1
62072,Lithuania,Vilnius,54.6833,25.3167,Europe/Vilnius,2025-04-01 12:30:00,11.1,Sunny,20.5,21,...,2.0,23.7,344.10,42.0,12.395,1.665,25.530,29.045,2,3
34652,Macedonia,Skopje,42.0000,21.4333,Europe/Skopje,2024-11-11 10:00:00,5.2,Light rain,3.6,88,...,0.9,10.8,403.30,34.0,19.980,12.210,34.780,37.000,2,3
33761,Belize,Belmopan,17.2500,-88.7667,America/Belize,2024-11-07 03:00:00,21.0,Clear,4.0,217,...,0.0,8.3,253.45,29.0,0.370,0.370,14.245,21.090,1,2
83159,Belgium,S Gravenjansdijk,51.2500,3.6333,Europe/Brussels,2025-07-18 10:30:00,24.1,Overcast,5.8,114,...,2.3,6.6,399.60,13.0,83.435,12.025,37.740,41.625,2,4
28676,Portugal,Lisbon,38.7167,-9.1333,Europe/Lisbon,2024-10-11 10:45:00,20.3,Partly cloudy,27.0,175,...,0.4,41.4,251.60,50.0,19.240,3.700,12.950,19.795,1,2
27466,Lithuania,Vilnius,54.6833,25.3167,Europe/Vilnius,2024-10-05 12:30:00,10.2,Partly cloudy,10.8,54,...,4.0,12.4,312.65,45.0,5.920,0.740,5.365,7.770,1,1


In [693]:
df.shape

(136632, 27)

In [694]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 136632 entries, 0 to 136632
Data columns (total 27 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136632 non-null  object        
 1   location                      136632 non-null  object        
 2   latitude                      136632 non-null  float64       
 3   longitude                     136632 non-null  float64       
 4   timezone                      136632 non-null  object        
 5   last_updated                  136632 non-null  datetime64[ns]
 6   temperature_celsius           136632 non-null  float64       
 7   condition_text                136632 non-null  object        
 8   wind_kph                      136632 non-null  float64       
 9   wind_degree                   136632 non-null  int64         
 10  wind_direction                136632 non-null  object        
 11  pressure_mb       

In [695]:
df.columns

Index(['country', 'location', 'latitude', 'longitude', 'timezone',
       'last_updated', 'temperature_celsius', 'condition_text', 'wind_kph',
       'wind_degree', 'wind_direction', 'pressure_mb', 'precip_mm', 'humidity',
       'cloud', 'feels_like_celsius', 'visibility_km', 'uv_index', 'gust_kph',
       'air_quality_Carbon_Monoxide', 'air_quality_Ozone',
       'air_quality_Nitrogen_dioxide', 'air_quality_Sulphur_dioxide',
       'air_quality_PM2.5', 'air_quality_PM10', 'epa_aqi', 'daqi'],
      dtype='object')

In [696]:
df.head()

,country,location,latitude,longitude,timezone,last_updated,temperature_celsius,condition_text,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.0,103.0,1.1,0.2,8.4,26.6,1,1
1,Albania,Tirana,41.33,19.82,Europe/Tirane,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,320,...,5.0,18.4,193.6,97.3,0.9,0.1,1.1,2.0,1,1
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,2024-05-16 09:45:00,23.0,Sunny,15.1,280,...,5.0,22.3,540.7,12.2,65.1,13.4,10.4,18.4,1,1
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,2024-05-16 10:45:00,6.3,Light drizzle,11.9,215,...,2.0,13.7,170.2,64.4,1.6,0.2,0.7,0.9,1,1
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,150,...,8.0,20.2,2964.0,19.0,72.7,31.5,183.4,262.3,5,10


In [697]:
df.dtypes

country                                 object
location                                object
latitude                               float64
longitude                              float64
timezone                                object
last_updated                    datetime64[ns]
temperature_celsius                    float64
condition_text                          object
wind_kph                               float64
wind_degree                              int64
wind_direction                          object
pressure_mb                              int64
precip_mm                              float64
humidity                                 int64
cloud                                    int64
feels_like_celsius                     float64
visibility_km                          float64
uv_index                               float64
gust_kph                               float64
air_quality_Carbon_Monoxide            float64
air_quality_Ozone                      float64
air_quality_N

## Rename Columns
- temperature_celsius = temp_c
- air_quality_PM2.5 = pm_2.5
- air-quality_PM10 = pm_10
- air_quality_carbon_monoxide = co
- air_quality_ozone = o3
- air_quality_nitrogen_dioxide = no2
- air_qality_sulphur_dioxide = so2


In [698]:
df.rename(columns={
    'air_quality_Carbon_Monoxide': 'co',
    'air_quality_Ozone': 'o3',
    'air_quality_Nitrogen_dioxide': 'no2',
    'air_quality_Sulphur_dioxide': 'so2',
    'air_quality_PM2.5': 'pm_2.5',
    'air_quality_PM10': 'pm_10', 
    'temperature_celsius': 'temp_c'
    },inplace=True)

C:\Users\ma lab\AppData\Local\Temp\ipykernel_12552\2308090726.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.rename(columns={


## Saving cleaned csv file

In [699]:
df.to_csv('weather_cleaned.csv', index=False)

In [700]:
df_new = pd.read_csv('weather_cleaned.csv')

In [701]:
df.head(1)

,country,location,latitude,longitude,timezone,last_updated,temp_c,condition_text,wind_kph,wind_degree,...,uv_index,gust_kph,co,o3,no2,so2,pm_2.5,pm_10,epa_aqi,daqi
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.0,103.0,1.1,0.2,8.4,26.6,1,1


In [702]:
df_new.dtypes

country                object
location               object
latitude              float64
longitude             float64
timezone               object
last_updated           object
temp_c                float64
condition_text         object
wind_kph              float64
wind_degree             int64
wind_direction         object
pressure_mb             int64
precip_mm             float64
humidity                int64
cloud                   int64
feels_like_celsius    float64
visibility_km         float64
uv_index              float64
gust_kph              float64
co                    float64
o3                    float64
no2                   float64
so2                   float64
pm_2.5                float64
pm_10                 float64
epa_aqi                 int64
daqi                    int64
dtype: object

# Summary Statistics

- mean
- median
- mode
- standard deviation

## Mean

In [703]:
print(df.select_dtypes(include='number').mean())
# to find mean of numerical columns

latitude                19.211176
longitude               21.956123
temp_c                  21.268111
wind_kph                12.897108
wind_degree            168.797280
pressure_mb           1014.064297
precip_mm                0.134169
humidity                66.661726
cloud                   39.976338
feels_like_celsius      22.120121
visibility_km            9.507446
uv_index                 3.313791
gust_kph                18.242809
co                     459.169180
o3                      57.998222
no2                     14.974119
so2                     10.315115
pm_2.5                  24.149295
pm_10                   48.293397
epa_aqi                  1.697428
daqi                     2.601279
dtype: float64


## Median

In [704]:
print(df.select_dtypes(include='number').median())
#to find median of numerical columns

latitude                17.2500
longitude               23.2361
temp_c                  23.9000
wind_kph                10.8000
wind_degree            161.0000
pressure_mb           1014.0000
precip_mm                0.0000
humidity                72.0000
cloud                   30.0000
feels_like_celsius      25.1000
visibility_km           10.0000
uv_index                 1.9000
gust_kph                15.3000
co                     293.7000
o3                      55.0000
no2                      5.6500
so2                      2.4050
pm_2.5                  14.1500
pm_10                   19.9800
epa_aqi                  1.0000
daqi                     2.0000
dtype: float64


## Mode

In [705]:
c=df.select_dtypes(include='object')
for i in c:
    print(i,df[i].mode())
    print()
    # to find mode of catogarical column

country 0    Bulgaria
Name: country, dtype: object

location 0            Accra
1            Amman
2           Asmara
3          Baghdad
4             Bern
5        Bujumbura
6            Dakar
7            Kabul
8             Kyiv
9           Malabo
10       N'djamena
11           Sanaa
12            Suva
13        Tashkent
14           Tokyo
15        Valletta
16    Vatican City
17          Warsaw
Name: location, dtype: object

timezone 0    Asia/Bangkok
Name: timezone, dtype: object

condition_text 0    Partly cloudy
Name: condition_text, dtype: object

wind_direction 0    E
Name: wind_direction, dtype: object



## Standard Deviation

In [706]:
print(df.select_dtypes(include='number').std())

latitude               24.416057
longitude              65.787067
temp_c                  9.687157
wind_kph               11.587170
wind_degree           103.752702
pressure_mb            10.327589
precip_mm               0.562252
humidity               23.871459
cloud                  34.096311
feels_like_celsius     11.590592
visibility_km           2.691788
uv_index                3.548004
gust_kph               13.793096
co                    755.709390
o3                     30.711697
no2                    23.649746
so2                    35.505696
pm_2.5                 36.761319
pm_10                 148.886879
epa_aqi                 0.939249
daqi                    2.440300
dtype: float64


In [707]:
stats_df = df.select_dtypes(include = 'number').agg (['mean','median','std']).T
stats_df

,mean,median,std
latitude,19.211176,17.2500,24.416057
longitude,21.956123,23.2361,65.787067
temp_c,21.268111,23.9000,9.687157
wind_kph,12.897108,10.8000,11.587170
wind_degree,168.797280,161.0000,103.752702
pressure_mb,1014.064297,1014.0000,10.327589
precip_mm,0.134169,0.0000,0.562252
humidity,66.661726,72.0000,23.871459
cloud,39.976338,30.0000,34.096311
feels_like_celsius,22.120121,25.1000,11.590592


In [710]:
 df.describe()

,latitude,longitude,last_updated,temp_c,wind_kph,wind_degree,pressure_mb,precip_mm,humidity,cloud,...,uv_index,gust_kph,co,o3,no2,so2,pm_2.5,pm_10,epa_aqi,daqi
count,136632.000000,136632.000000,136632,136632.000000,136632.000000,136632.000000,136632.000000,136632.000000,136632.000000,136632.000000,...,136632.000000,136632.000000,136632.00000,136632.000000,136632.000000,136632.000000,136632.000000,136632.000000,136632.000000,136632.000000
mean,19.211176,21.956123,2025-05-03 09:03:41.923854080,21.268111,12.897108,168.797280,1014.064297,0.134169,66.661726,39.976338,...,3.313791,18.242809,459.16918,57.998222,14.974119,10.315115,24.149295,48.293397,1.697428,2.601279
min,-41.300000,-175.200000,2024-05-16 01:45:00,-29.800000,3.600000,1.000000,947.000000,0.000000,2.000000,0.000000,...,0.000000,3.600000,-9999.00000,0.000000,0.000000,-9999.000000,0.168000,-1848.150000,1.000000,1.000000
25%,4.050300,-6.836100,2024-11-09 05:00:00,15.900000,6.100000,80.000000,1010.000000,0.000000,51.000000,0.000000,...,0.100000,10.200000,201.85000,38.000000,1.687000,1.110000,7.050000,9.990000,1.000000,1.000000
50%,17.250000,23.236100,2025-05-03 11:00:00,23.900000,10.800000,161.000000,1014.000000,0.000000,72.000000,30.000000,...,1.900000,15.300000,293.70000,55.000000,5.650000,2.405000,14.150000,19.980000,1.000000,2.000000
75%,40.400000,49.882200,2025-10-25 14:22:30,28.000000,17.600000,256.000000,1018.000000,0.030000,86.000000,75.000000,...,6.000000,24.200000,458.80000,74.000000,17.390000,8.325000,27.750000,41.440000,2.000000,3.000000
max,64.150000,179.220000,2026-04-20 19:30:00,49.200000,2963.200000,360.000000,3006.000000,42.240000,100.000000,100.000000,...,16.300000,2970.400000,38879.39800,480.700000,427.700000,521.330000,1614.100000,6037.290000,6.000000,10.000000
std,24.416057,65.787067,NaN,9.687157,11.587170,103.752702,10.327589,0.562252,23.871459,34.096311,...,3.548004,13.793096,755.70939,30.711697,23.649746,35.505696,36.761319,148.886879,0.939249,2.440300


In [711]:
df.describe().T

,count,mean,min,25%,50%,75%,max,std
latitude,136632.0,19.211176,-41.3,4.0503,17.25,40.4,64.15,24.416057
longitude,136632.0,21.956123,-175.2,-6.8361,23.2361,49.8822,179.22,65.787067
last_updated,136632,2025-05-03 09:03:41.923854080,2024-05-16 01:45:00,2024-11-09 05:00:00,2025-05-03 11:00:00,2025-10-25 14:22:30,2026-04-20 19:30:00,NaN
temp_c,136632.0,21.268111,-29.8,15.9,23.9,28.0,49.2,9.687157
wind_kph,136632.0,12.897108,3.6,6.1,10.8,17.6,2963.2,11.58717
wind_degree,136632.0,168.79728,1.0,80.0,161.0,256.0,360.0,103.752702
pressure_mb,136632.0,1014.064297,947.0,1010.0,1014.0,1018.0,3006.0,10.327589
precip_mm,136632.0,0.134169,0.0,0.0,0.0,0.03,42.24,0.562252
humidity,136632.0,66.661726,2.0,51.0,72.0,86.0,100.0,23.871459
cloud,136632.0,39.976338,0.0,0.0,30.0,75.0,100.0,34.096311


## Informations :-

### same mean and median

##### Same:
- pressure_mb
- epa_aqi
- daqi
- presip_mm
- visibility_km
- TOTAL = 5

##### Almost Same :-

- feels_like_same
- temp_c
- latitude
- longitude
- wind_kph
- gust_kph
- o3
- TOTAL = 7

##### Defferent :-

- co
- pm_2.5
- pm_10
- wind_degree
- humidity
- cloud
- uv_index
- so2
- no2
- TOTAL = 9

In [712]:
df.describe(include=object)

,country,location,timezone,condition_text,wind_direction
count,136632,136632,136632,136632,136632
unique,210,257,199,49,16
top,Bulgaria,Kabul,Asia/Bangkok,Partly cloudy,E
freq,1621,703,2516,39376,12995


## Informations :-

- the dataset is compleate with no missing categorical values.
- there is high geographical diversity (200+ countries and locations). 
- weather conditions are repititive, with "party cloudy" being the most common.
- wind direction shows a pattern, with east being the most frequent.
- some locations (like kabul) appear more often, suggesting the dataset may be unevenly distributed.